config/deployment.yaml

```diff
@@ -62,6 +65,10 @@ spec:
       - name: fulcio-config
         configMap:
           name: fulcio-config
+      - name: fulcio-secret
+        secret:
+          secretName: fulcio-secret
+          optional: true
```

Comment:
^^ @mattmoor

Reply:
That should make it so that it doesn't block starting it. The one interesting bit is what happens if you do want to specify the secret and it is not there yet but you'd like it to actually block. IIRC though, we watch that mount point so I _think_ it will pick it up once it becomes available. @nsmith5 check me out :)

Reply:
Looks like `--fileca-watch` flag defaults to true, so I think thats correct @vaikas

Reply:
`fileca` will never start without a valid key-pair. We never want to boot up the server if we can't serve requests so I wrote it to try and load an initial key-pair and bail hard if the doesn't work (ref https://github.com/sigstore/fulcio/blob/main/pkg/ca/fileca/fileca.go#L48)

I don't think I know what Kubernetes would do in this case. My hope would be a pod crashloop until the secret mounts, but I don't know the `optional` spec well enough to know that is the case. Once the pod from this deployment is scheduled and the secret wasn't available at schedule time, does that mean it will never get that secret volume mounted by pods scheduled later when the secret exists _will_ get it? Or does it mean that Kubernetes will try to mount it even to an existing pod once the secret does exist?

Reply:
The watch behaviour is also pretty paranoid about putting the server in a state where it can't serve requests so it looks for updates, but if, for instance, the cert doesn't match the key or the key can't be decrypted with the password it will just keep using the existing certs and ignore the updates from the filesystem

Reply:
I think the real issue here is that we should not require the pod to crashloop if the secret is missing, unless the `fileca` is being specified, because in those cases, there's no spoon (err, secret).
That was the thinking behind having the secret be optionally mounted, and then depending on the flags specified, we'd do something different.
When secrets change after they've been mounted, they will eventually be updated on the volume mount. But the question about if theres' no secret at startup time and then it shows up if that will still hold true is a good one, and for that I do not know the answer. I'd expect that an empty volume is mounted and when the secret shows up it will be populated, but that's just my expectation as a user.

Reply:
Oops, I read your reply incorrectly :) Yes, if the secret is missing, the pod will crashloop because there's no secret. In my head I read that the pod won't start. So, yes.
But the question is still if non-existent secret at startup time will still get updated once it shows up. And I do not know answer to that.

Reply:
So the default config is using `googleca` - so is this really any concern? If the files are missing using googleca, then .. no issue right?

Reply:
see https://github.com/sigstore/fulcio/blob/main/config/fulcio-config.yaml#L61

Reply:
> The one interesting bit is what happens if you do want to specify the secret and it is not there yet but you'd like it to actually block

I think what we need to resolve this via Dr. Empirical is:
1. Create a Pod with an optional secret volume and an image that crash loops if the secret isn't there.
2. Create the secret and see if it stops crash looping.

I don't believe K8s will mount an optional secret into a running pod, but it is plausible that it might when restarting containers.  An interesting corollary here is: Does K8s refresh secrets projected into env vars when a container crashes?

Reply:
We can totally do an experiment, my only thought is that if the behaviour changes from version to version if it's not explicitly specified. But mosdef worth testing.

Reply:
Ran Matt's suggestion, and things appear to work as we expect:

1. Create a Pod with an optional secret volume and an image that crash loops if the secret isn't there:

```
cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: boop-deployment
  labels:
    app: boop
spec:
  replicas: 1
  selector:
    matchLabels:
      app: boop
  template:
    metadata:
      labels:
        app: boop
    spec:
      containers:
      - name: boop
        image: bash
        command:
        - "bash"
        - "-c"
        - "[[ -f /boop/bop.txt ]] || (echo bad && exit 1) && (echo good && while true; do sleep 1; done)"
        volumeMounts:
        - name: boop-secret
          mountPath: /boop
          readOnly: true
      volumes:
      - name: boop-secret
        secret:
          secretName: boop-secret
          optional: true
EOF
```

```
$ k get pods
NAME                               READY   STATUS             RESTARTS   AGE
boop-deployment-5c48fd856b-tv9lz   0/1     CrashLoopBackOff   2          33s
```

2. Create the secret and see if it stops crash looping.

```
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: boop-secret
data:
  bop.txt: $(echo boop | base64)
EOF
```

```
$ k get pods
NAME                               READY   STATUS    RESTARTS   AGE
boop-deployment-5c48fd856b-tv9lz   1/1     Running   3          49s
```

yay?

Reply:
@vaikas @nsmith5 Did you have anything else on this before we merge?

Reply:
Nope looks good from me! Good work @jdolitsky :D

Reply:
Thanks for checking in with Dr. Empirical 😆

Reply:
No. I think it might behoove us to file an issue in k8s to document the behaviour in case they reserve the right to change this behaviour. But I'm happy with the change for sure.